### ISSUE TITLE
Binning for mutual info calculation

### ISSUE THREAD

**Issue by pmorerio** [ravidziv/IDNNs#12]
Hi, I have been running your code with the option `-activation_function=1`, i.e. with ReLUs according to the README. However, I noticed that binning for mutual info calculation is performed in between -1 and 1 _independent_ of the activation function.
https://github.com/ravidziv/IDNNs/blob/c4abb1dad4fbb262315eb8b96eb85dc5c3e98e5c/idnns/information/information_process.py#L183

**Comment by pmorerio** [comment]
Hi again. I changed the line of code above in order to have a bin interval which should better reflect the asymmetric  range of values of ReLUs (approximately; I may be missing some bigger values).
```
bins = np.linspace(0, 10, num_of_bins)
```
Since this increases the range, I also considered more bins with `-num_of_bins=100`.

This is what I get.

![figure](https://user-images.githubusercontent.com/6783835/36599031-54839334-18ae-11e8-80bc-f697856b7c79.jpg)

**Comment by gladiator8072** [comment]
i think somewhat you have similar figure compared with the reviewing paper
https://openreview.net/forum?id=ry_WPG-A-
see Figure 1. (B)
while there's few comments below and worth to check

**Comment by pmorerio** [comment]
@gladiator8072 yes indeed. Thanks. I have read the paper and the comments. I am inclined to think that the compression phase is to be ascribed to the saturation of the `tanh` and to this wrong binning for the `ReLU`. What do you think?

**Comment by gladiator8072** [comment]
not_so_sure, except bin range, the number of bins might cause some noise effects and also relu is unbound on positive side, implies that the range of such direction is not fixed. 
actually they also defends the relu result in the comments: https://goo.gl/VrRm5W. which shows the relu also has compression phase. 



please let me know you find some secret :D

**Comment by pmorerio** [comment]
Thanks, I read that document. Naftali Tishby and @ravidziv  state that the authors of the ICLR paper 
>  don’t know how to estimate mutual information correctly

However, they do not explain how such information should be estimated in practice, and the code they provide here certainly does a wrong estimate for `ReLUs`. It would be great to have an update of  this repo which replicates the results they show. This would  clarify everything and shed some light.

**Comment by pmorerio** [comment]
@gladiator8072 The code for that ICLR paper is here https://github.com/pmorerio/ibsgd

**Comment by gladiator8072** [comment]
awesome! i'm going to study this!!
